gtsam/discrete/DecisionTree-inl.h

```diff
@@ -75,15 +77,38 @@ namespace gtsam {
       return (q.isLeaf() && q.sameLeaf(*this));
     }
 
+    /// @{
+    /// SFINAE methods for proper substitution.
+    /** equality for integral types. */
+    template <typename T = Y>
+    typename std::enable_if<std::is_integral<T>::value, bool>::type
+    equals(const T& a, const T& b, double tol) const {
+      return std::abs(double(a - b)) < tol;
+    }
+    /** equality for boost::shared_ptr types. */
+    template <typename T = Y>
+    typename std::enable_if<boost::has_dereference<T>::value, bool>::type
+    equals(const T& a, const T& b, double tol) const {
+      return traits<typename T::element_type>::Equals(*a, *b, tol);
+    }
+    /** equality for all other types. */
+    template <typename T = Y>
+    typename std::enable_if<!boost::has_dereference<T>::value && !std::is_integral<T>::value, bool>::type
+    equals(const Y& a, const Y& b, double tol) const {
+      return traits<Y>::Equals(a, b, tol);
+    }
+    /// @}
+
     /** equality up to tolerance */
     bool equals(const Node& q, double tol) const override {
-      const Leaf* other = dynamic_cast<const Leaf*> (&q);
+      const Leaf* other = dynamic_cast<const Leaf*>(&q);
       if (!other) return false;
-      return std::abs(double(this->constant_ - other->constant_)) < tol;
+      return this->equals<Y>(this->constant_, other->constant_, tol);
```

Comment:
What happened to passing a compare method? I had advised to keep DT generic, independent of GTSAM :-)

Reply:
Yeah I had a question about that and just wanted something working for now. This `equals` is in the `Leaf` class, so we would have to add the comparison method for all node types. Is that intended?